src/parameterizations/vertical/MOM_vert_friction.F90

```diff
@@ -643,341 +648,485 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
+      zDS = 0.0
+      stress = dt_Rho0 * forces%taux(I,j)
+      do k=1,nz
+        h_a = 0.5 * (h(i,j,k) + h(i+1,j,k)) + h_neglect
+        hfr = 1.0 ; if ((zDS+h_a) > Hmix) hfr = (Hmix - zDS) / h_a
+        u(I,j,k) = u(I,j,k) + I_Hmix * hfr * stress
+        if (associated(ADp%du_dt_str)) ADp%du_dt_str(i,J,k) = (I_Hmix * hfr * stress) * Idt
+        zDS = zDS + h_a ; if (zDS >= Hmix) exit
+      enddo
     endif ; enddo ; enddo
+  else
+    do j=G%jsc,G%jec ; do I=Isq,Ieq
+      surface_stress(I,j) = dt_Rho0 * (G%mask2dCu(I,j)*forces%taux(I,j))
+    enddo ; enddo
+  endif
 
-    ! back substitute to solve for the new velocities
-    ! u_k = d'_k - c'_k x_(k+1)
-    do k=nz-1,1,-1 ; do I=Isq,Ieq ; if (do_i(I)) then
-      u(I,j,k) = u(I,j,k) + c1(I,k+1) * u(I,j,k+1)
-    endif ; enddo ; enddo ! i and k loops
+  ! perform forward elimination on the tridiagonal system
+  !
+  ! denote the diagonal of the system as b_k, the subdiagonal as a_k
+  ! and the superdiagonal as c_k. The right-hand side terms are d_k.
+  !
+  ! ignoring the Rayleigh drag contribution,
+  ! we have a_k = -dt * a_u(k)
+  !         b_k = h_u(k) + dt * (a_u(k) + a_u(k+1))
+  !         c_k = -dt * a_u(k+1)
+  !
+  ! for forward elimination, we want to:
+  ! calculate c'_k = - c_k                / (b_k + a_k c'_(k-1))
+  ! and       d'_k = (d_k - a_k d'_(k-1)) / (b_k + a_k c'_(k-1))
+  ! where c'_1 = c_1/b_1 and d'_1 = d_1/b_1
+  !
+  ! This form is mathematically equivalent to Thomas' tridiagonal matrix algorithm, but it
+  ! does not suffer from the acute sensitivity to truncation errors of the Thomas algorithm
+  ! because it involves no subtraction, as discussed by Schopf & Loughe, MWR, 1995.
+  !
+  ! b1 is the denominator term 1 / (b_k + a_k c'_(k-1))
+  ! b_denom_1 is (b_k + a_k + c_k) - a_k(1 - c'_(k-1))
+  !            = (b_k + c_k + c'_(k-1))
+  ! this is done so that d1 = b1 * b_denom_1 = 1 - c'_(k-1)
+  ! c1(k) is -c'_(k - 1)
+  ! and the right-hand-side is destructively updated to be d'_k
+
+  if (allocated(visc%Ray_u)) then
+    do j=G%jsc,G%jec ; do I=Isq,Ieq
+      Ray(I,j) = visc%Ray_u(I,j,1)
+    enddo ; enddo
+  else
+    do j=G%jsc,G%jec ; do I=Isq,Ieq
+      Ray(I,j) = 0.
+    enddo ; enddo
+  endif
+
+  do j=G%isc,G%jec ; do I=Isq,Ieq ; if (G%mask2dCu(I,j) > 0.) then
+    b_denom_1 = CS%h_u(I,j,1) + dt * (Ray(I,j) + CS%a_u(I,j,1))
+    b1(I,j) = 1.0 / (b_denom_1 + dt*CS%a_u(I,j,2))
+    d1(I,j) = b_denom_1 * b1(I,j)
+    u(I,j,1) = b1(I,j) * (CS%h_u(I,j,1) * u(I,j,1) + surface_stress(I,j))
+  endif ; enddo ; enddo
+
+  if (associated(ADp%du_dt_str)) then
+    do j=G%isc,G%jec ; do I=Isq,Ieq ; if (G%mask2dCu(I,j) > 0.) then
+      ADp%du_dt_str(I,j,1) = b1(I,j) * ( &
+            CS%h_u(I,j,1) * ADp%du_dt_str(I,j,1) + surface_stress(I,j) * Idt &
+          )
+    endif ; enddo ; enddo
+  endif
+
+  do k=2,nz
+    if (allocated(visc%Ray_u)) then
+      do j=G%jsc,G%jec ; do I=Isq,Ieq
+        Ray(I,j) = visc%Ray_u(I,j,k)
+      enddo ; enddo
+    else
```

Comment:
Unless I am missing something, this `else` block seems unnecessary, as `Ray(:,:)` was already set to zero at line 741 above.  Should the `else` block here and at line 999 be removed?